helix-core/src/main/java/org/apache/helix/tools/ClusterVerifiers/BestPossibleExternalViewVerifier.java

```diff
@@ -274,8 +276,18 @@ protected synchronized boolean verifyState() {
 
       // Filter resources if requested
       if (_resources != null && !_resources.isEmpty()) {
-        idealStates.keySet().retainAll(_resources);
-        extViews.keySet().retainAll(_resources);
+        Set<String> wagedResource = _resources.stream().filter(
+                resourceEntry -> WagedValidationUtil.isWagedEnabled(idealStates.get(resourceEntry)))
+            .collect(Collectors.toSet());
+        if (wagedResource.isEmpty()) {
+          // If no waged-enabled resources are found, retain only the provided resources
+          idealStates.keySet().retainAll(_resources);
+          extViews.keySet().retainAll(_resources);
+        } else {
+          // If any of the resources is a waged resource, we need to verify all resources
+          // because the best possible state calculation is done for all resources
+          _resources.clear();
```

Comment:
Let's avoid modifying the state of `_resources`. When a user specifies resources, they only want to check that those resources have converged with the BPS. That being said, if `_resources` contains a WAGED resource we should not filter WAGED resources out from the `idealStates` or `extViews`(You can filter non-waged resources that are not in `_resources` as an optimization)

When we pass all WAGED resources to `calcBestPossState`, we should only actually validate and compare the state for `_resources` instead of for the whole cluster.